src/main.rs

```diff
@@ -330,13 +332,84 @@ extern "C" fn primary_init(_hart_id: u64, _fdt_addr: u64) -> CpuParams {
 #[no_mangle]
 extern "C" fn primary_main() {}
 
-/// Rust Entry point for the bootstrap CPU. Bootstraps salus, starts secondary CPUs and builds the
-/// HOST VM.
+#[derive(Debug)]
+enum RequiredCpuFeature {
+    Aia,
+    Sstc,
+}
+
+#[derive(Debug)]
+enum RequiredDeviceProbe {
+    Imsic(drivers::imsic::ImsicError),
+    Pci(drivers::pci::PciError),
+    Reset(drivers::reset::Error),
+    Uart(drivers::uart::Error),
+}
+
+impl Display for RequiredDeviceProbe {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        use RequiredDeviceProbe::*;
+        match self {
+            Imsic(e) => write!(f, "IMSIC: {:?}", e),
+            Pci(e) => write!(f, "PCI: {:?}", e),
+            Reset(e) => write!(f, "Reset: {:?}", e),
+            Uart(e) => write!(f, "UART: {:?}", e),
+        }
+    }
+}
```

Comment:
I think this is a bit too boiler-plate-y for something that might change in the future. I would rather push something like this in the drivers infrastracture, so that it becomes a standard probe return, rather than forcing it here.

Reply:
The other option was to add 4 different errors to the main `Error` enum and I thought this was less intrusive.

Reply:
Yes, we need to check in the future then that other device rpobes use that error code.